asmcomp/reg.ml

```diff
@@ -200,3 +200,14 @@ let set_of_array v =
   | n -> let rec add_all i =
            if i >= n then Set.empty else Set.add v.(i) (add_all(i+1))
          in add_all 0
+
+let visit_generation = ref 1
+
+let mark_visited r =
+  r.visited <- !visit_generation
+
+let is_visited r =
+  r.visited = !visit_generation
+
+let clear_visited_marks () =
+  incr visit_generation
```

Comment:
How (un)likely is it for `visit_generation` to wrap around to 0, on a 32-bit machine?

Reply:
Hmm. I don't think doing 2^32 register visits within compilation of a single function is plausible (and even if it was, the penalty is merely slightly-worse codegen).

However, it's almost believable that you could overflow this in a large compilation unit with many functions, which could mean that newly-created registers seem visited by colliding with visit_generation. So, I've changed the creation logic so that newly-created registers get `!visit_generation - 1` as their `visited` field, so no register can incorrectly seem visited unless a positive multiple of 2^32 visits have happened since it was created.

Reply:
On second thoughts, I don't think it's believable to ever overflow this. I think that compilation units big enough to overflow are not representable on 32-bit address spaces. Still, I prefer the new versions.

Reply:
> I've changed the creation logic so that newly-created registers get !visit_generation - 1 as their visited field

I think you can achieve the same result by resetting `visit_generation` to 1 in `Reg.reset`.  That's the first idea that crossed my mind, and maybe it's a little more intuitive.

Both solutions make me hesitate because of the precolored registers that represent hardware registers.  Those registers are created once and for all, not re-created for every function, so their "visited" field is unpredictable.

On second thought, there are no preference edges from a precolored register, so it doesn't matter whether they are not visited  (because their `visited` field is big enough) or visited (their  `visited `field will be updated, but no preference edge will be followed).

So, I think your fix is correct, and so is mine (resetting `visit_generation` as part of `Reg.reset`).

Reply:
Could I suggest we add some of this reasoning (e.g. about pre-coloured registers) as a comment in the code?

Reply:
I didn't think of using `Reg.reset` here. I think that solution is better, I've changed to use that instead.

I found the "hardware registers don't have preference edges" reasoning a bit subtle, so instead I've opted to explicitly reset the visit marks of hw registers during Reg.reset.